### PR TITLE
Complete decoding of cmovcc and jcc

### PR DESCRIPTION
**Pull request by aengelke**
This adds a complete decoding of all CMOVcc and handling of Jcc instructions.

### CODE REVIEW

File: src/decode.c [hunk 1]
```diff
@@ -394,25 +394,23 @@ DBB* dbrew_decode(Rewriter* r, uint64_t f)
...
+            case 0x47:
+            case 0x48:
+            case 0x49:
+            case 0x4A:
+            case 0x4B:
+            case 0x4C:
+            case 0x4D:
+            case 0x4E:
+            case 0x4F:
+                it = IT_CMOVO + (opc2 & 0xf);
```
Comment (weidendo) [inline src/decode.c]: Please do not create a dependance between IT_* enum and opcode numbers, this just asks for bugs.
Just leave it as before, and extend it by the new cases. And please keep/add mnenomics comments.

File: src/decode.c [hunk 2]
```diff
@@ -496,24 +494,26 @@ DBB* dbrew_decode(Rewriter* r, uint64_t f)
...
                 off += 4;
-                if      (opc2 == 0x84) it = IT_JE;
-                else if (opc2 == 0x85) it = IT_JNE;
-                else if (opc2 == 0x8A) it = IT_JP;
-                else if (opc2 == 0x8C) it = IT_JL;
-                else if (opc2 == 0x8D) it = IT_JGE;
-                else if (opc2 == 0x8E) it = IT_JLE;
-                else if (opc2 == 0x8F) it = IT_JG;
-                else assert(0);
+                it = IT_JO + (opc2 & 0xf);
```
Comment (weidendo) [inline src/decode.c]: See comment above. Better use a switch here.

File: src/decode.c [hunk 3]
```diff
@@ -746,24 +746,26 @@ DBB* dbrew_decode(Rewriter* r, uint64_t f)
...
             off += 1;
-            if      (opc == 0x74) it = IT_JE;
-            else if (opc == 0x75) it = IT_JNE;
-            else if (opc == 0x7A) it = IT_JP;
-            else if (opc == 0x7C) it = IT_JL;
-            else if (opc == 0x7D) it = IT_JGE;
-            else if (opc == 0x7E) it = IT_JLE;
-            else if (opc == 0x7F) it = IT_JG;
-            else assert(0);
+            it = IT_JO + (opc & 0xf);
```
Comment (weidendo) [inline src/decode.c]: Dito.

**Comment by weidendo** [comment]
I do not like this creating an unneeded dependance between IT_* enums and opcode numbers.
Please keep this as it was. And to test our current test infrastructure: Before merging this, can we have  test cases checking correct decoding?

This should be followed by complete support for the instructions (as is done with currently supported instrucutions): emulation and capturing. Tests for jump instructions need to test different paths, and for capturing also have cases where the condition is known/static and unknown/dynamic.

**Comment by aengelke** [comment]
This should be changed now. Travis build fails due to addresses in the output :disappointed:

**Comment by aengelke** [comment]
Now, the jump instructions also get emulated. Also, I added more test cases.

**Comment by weidendo** [comment]
Cool!

We need to get rid of these addresses in the output.
From the travis output:

  Decoding BB test ...
                  test:  48 83 c7 02           add     $0x2,%rdi
!               test+4:  7a 03                 jp      $0x400a2f

This needs the instruction printer to be able to output named addresses with offsets
(here: test+XX). We must pass on the Rewriter within the printer, and use prettyAddress().
This also should be done when printing constant displacement of operands accessing
memory.

  Emulate 'test: add $0x2,%rdi'
! Emulate 'test+4: jp $0x400a2f'

Should work when the above works.

! Decoding BB 0x400a2f ...

This needs to be "Decoding BB test+XX ...".
Not sure what's the best way to get the size of the code of a function?
Even how to do this explicit just for tests is not clear to me. Any ideas?

!             0x400a2f:  b8 01 00 00 00        mov     $0x1,%eax

Should work when the above works.

**Comment by weidendo** [comment]
Current master should be in a state to make the tests work.
Does the PR have to be rebased for that?

Current master also has a op-je.s test case with run options to test the different branches.

BTW, I used "je" instead of "jz" (both are allowed by Intel and generate the same machine code) in the printer to stay as compatible as possible to objdump assembly output. I would do the same for the
other conditions in this PR.

**Comment by aengelke** [comment]
Should be ready to merge now.

**Pull request merged**